community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseFactoryState.java

```diff
@@ -37,29 +37,32 @@
 
 public class GraphDatabaseFactoryState
 {
+    // Keep these fields volatile or equivalent because of this scenario:
+    // - one thread creates a GraphDatabaseFactory (including state)
+    // - this factory will potentially be handed over to other threads, which will create databases
     private final List<Class<?>> settingsClasses;
     private final List<KernelExtensionFactory<?>> kernelExtensions;
-    private Monitors monitors;
-    private LogProvider userLogProvider;
+    private volatile Monitors monitors;
+    private volatile LogProvider userLogProvider;
     private final Map<String,URLAccessRule> urlAccessRules;
 
     public GraphDatabaseFactoryState()
     {
-        settingsClasses = new ArrayList<>();
+        settingsClasses = new CopyOnWriteArrayList<>();
         settingsClasses.add( GraphDatabaseSettings.class );
-        kernelExtensions = new ArrayList<>();
+        kernelExtensions = new CopyOnWriteArrayList<>();
```

Comment:
The `removeKernelExtensions` method now needs to be implemented differently, because you cannot remove from a `CopyOnWriteArrayList` via its iterator.